src/main/scala/extractor/SemanticCtx.scala

```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
 }
 
 object SemanticDBFileVisitor {
-  def apply[T](filePath: Path, projectPath: AbsolutePath, results: ConcurrentLinkedQueue[(String, T)], f: SemanticCtx => Iterable[(String, T)]) = {
+  def apply[T](filePath: Path, f: SemanticCtx => Result): Result = {
       try {
         val sdb = s.Database.parseFrom(Files.readAllBytes(filePath))
         val mdb = sdb.toDb(None)
-        val ctx = SemanticCtx(mdb, projectPath)
-        f(ctx).foreach(results.add)
+        val ctx = SemanticCtx(mdb)
+        val res = f(ctx)
         print(".")
+        res
       } catch {
         case NonFatal(e) =>
           val st = e.getStackTrace
           e.setStackTrace(st.take(10))
           e.printStackTrace()
+          Result(Set(), Seq(), Set(), Set())
       }
   }
 }
 
-abstract class SemanticDBWalker {
-  def run[T](f: SemanticCtx => Iterable[(String, T)]): ConcurrentLinkedQueue[(String, T)]
+abstract class TreeWalker {
```

Comment:
This seem like a function as well.
Moreover does not have any state so it can be a trait.